File: src/commands/remove.ts
import {
  removePackageSource,
  removeRepoSource,
  repoExists,
  listSources,
  getPackageInfo,
} from "../lib/git.js";
import {
  updateAgentsMd,
  updatePackageIndex,
  type PackageEntry,
  type RepoEntry,
} from "../lib/agents.js";
import { isRepoSpec } from "../lib/repo.js";
import { getFileModificationPermission } from "../lib/settings.js";
import { detectRegistry } from "../lib/registries/index.js";
import type { Registry } from "../types.js";

export interface RemoveOptions {
  cwd?: string;
}

/**
 * Remove source code for one or more packages or repositories
 */
export async function removeCommand(
  items: string[],
  options: RemoveOptions = {},
): Promise<void> {
  const cwd = options.cwd || process.cwd();
  let removed = 0;
  let notFound = 0;

  // Track packages and repos to update in sources.json
  const removedPackages: Array<{ name: string; registry: Registry }> = [];
  const removedRepos: string[] = [];

  for (const item of items) {
    // Check if it's a repo or package based on format
    const isRepo =
      isRepoSpec(item) || (item.includes("/") && !item.includes(":"));

    if (isRepo) {
      // Try to remove as repo
      // Convert formats like "vercel/vercel" to "github.com/vercel/vercel" if needed
      let displayName = item;
      if (item.split("/").length === 2 && !item.startsWith("http")) {
        displayName = `github.com/${item}`;
      }

      if (!repoExists(displayName, cwd)) {
        // Try the item as-is (might already be full path like github.com/owner/repo)
        if (repoExists(item, cwd)) {
          displayName = item;
        } else {
          console.log(`  ⚠ ${item} not found`);
          notFound++;
          continue;
        }
      }

      const success = await removeRepoSource(displayName, cwd);

      if (success) {
        console.log(`  ✓ Removed ${displayName}`);
        removed++;
        removedRepos.push(displayName);
      } else {
        console.log(`  ✗ Failed to remove ${displayName}`);
      }
    } else {
      // Remove as package - detect registry from prefix or default to npm
      const { registry, cleanSpec } = detectRegistry(item);

      // Find the package in sources
      let pkgInfo = await getPackageInfo(cleanSpec, cwd, registry);
      let actualRegistry = registry;

      if (!pkgInfo) {
        // Try other registries if default didn't work
        const registries: Registry[] = ["npm", "pypi", "crates"];
        for (const reg of registries) {
          if (reg !== registry) {
            pkgInfo = await getPackageInfo(cleanSpec, cwd, reg);
            if (pkgInfo) {
              actualRegistry = reg;
              break;
            }
          }
        }
      }

      if (!pkgInfo) {
        console.log(`  ⚠ ${cleanSpec} not found`);
        notFound++;
        continue;
      }

      const result = await removePackageSource(cleanSpec, cwd, actualRegistry);

      if (result.removed) {
        console.log(`  ✓ Removed ${cleanSpec} (${actualRegistry})`);
        if (result.repoRemoved) {
          console.log(`    → Also removed repo (no other packages use it)`);
        }
        removed++;
        removedPackages.push({ name: cleanSpec, registry: actualRegistry });
      } else {
        console.log(`  ✗ Failed to remove ${cleanSpec}`);
      }
    }
  }

  console.log(
    `\nRemoved ${removed} source(s)${notFound > 0 ? `, ${notFound} not found` : ""}`,
  );

  // Update sources.json with remaining sources
  if (removed > 0) {
    const sources = await listSources(cwd);

    // Filter out removed packages
    const remainingPackages: PackageEntry[] = sources.packages.filter(
      (p) =>
        !removedPackages.some(
          (rp) => rp.name === p.name && rp.registry === p.registry,
        ),
    );

    // Filter out removed repos
    const remainingRepos: RepoEntry[] = sources.repos.filter(
      (r) => !removedRepos.includes(r.name),
    );

    // Check if file modifications are allowed
    const canModifyFiles = await getFileModificationPermission(cwd);

    if (canModifyFiles) {
      const agentsUpdated = await updateAgentsMd(
        { packages: remainingPackages, repos: remainingRepos },
        cwd,
      );
      if (agentsUpdated) {
        const totalRemaining = remainingPackages.length + remainingRepos.length;
        if (totalRemaining === 0) {
          console.log("✓ Removed opensrc section from AGENTS.md");
        } else {
          console.log("✓ Updated AGENTS.md");
        }
      }
    } else {
      await updatePackageIndex(
        { packages: remainingPackages, repos: remainingRepos },
        cwd,
      );
    }
  }
}

File: packages/opensrc/cli/src/commands/remove.rs
use crate::core::cache::{
    get_package_info, list_sources, remove_package_source, remove_repo_source, write_sources,
};
use crate::core::error::{Error, Result};
use crate::core::registries::repo::{is_repo_spec, parse_repo_spec};
use crate::core::registries::{detect_registry, Registry};

pub fn run(items: &[String]) -> Result<()> {
    let mut removed = 0u32;
    let mut not_found = 0u32;
    let mut had_errors = false;

    let mut removed_packages: Vec<(String, Registry)> = Vec::new();
    let mut removed_repos: Vec<String> = Vec::new();

    for item in items {
        let is_repo = is_repo_spec(item) || (item.contains('/') && !item.contains(':'));

        if is_repo {
            let display_name = match parse_repo_spec(item) {
                Some(spec) => format!("{}/{}/{}", spec.host, spec.owner, spec.repo),
                None => {
                    println!("  ✗ Could not parse repo spec: {item}");
                    had_errors = true;
                    continue;
                }
            };

            match remove_repo_source(&display_name, None) {
                Ok(true) => {
                    println!("  ✓ Removed {display_name}");
                    removed += 1;
                    removed_repos.push(display_name);
                }
                Ok(false) => {
                    println!("  ⚠ {item} not found");
                    not_found += 1;
                }
                Err(e) => {
                    println!("  ✗ Error removing {item}: {e}");
                    had_errors = true;
                }
            }
        } else {
            let detected = detect_registry(item);
            let clean = &detected.clean_spec;
            let mut registry = detected.registry;

            let mut pkg_info = get_package_info(clean, registry)?;

            if pkg_info.is_none() {
                let registries = [Registry::Npm, Registry::PyPI, Registry::Crates];
                for reg in &registries {
                    if *reg != registry {
                        if let Some(info) = get_package_info(clean, *reg)? {
                            pkg_info = Some(info);
                            registry = *reg;
                            break;
                        }
                    }
                }
            }

            if pkg_info.is_none() {
                println!("  ⚠ {clean} not found");
                not_found += 1;
                continue;
            }

            match remove_package_source(clean, registry) {
                Ok((true, repo_removed)) => {
                    println!("  ✓ Removed {clean} ({registry})");
                    if repo_removed {
                        println!("    → Also removed repo (no other packages use it)");
                    }
                    removed += 1;
                    removed_packages.push((clean.clone(), registry));
                }
                Ok((false, _)) => {
                    println!("  ✗ Failed to remove {clean}");
                    had_errors = true;
                }
                Err(e) => {
                    println!("  ✗ Error removing {clean}: {e}");
                    had_errors = true;
                }
            }
        }
    }

    let nf_msg = if not_found > 0 {
        format!(", {not_found} not found")
    } else {
        String::new()
    };
    println!("\nRemoved {removed} source(s){nf_msg}");

    if removed > 0 {
        let (packages, repos) = list_sources()?;

        let remaining_packages: Vec<_> = packages
            .into_iter()
            .filter(|p| {
                !removed_packages
                    .iter()
                    .any(|(name, reg)| p.name == *name && p.registry == *reg)
            })
            .collect();

        let remaining_repos: Vec<_> = repos
            .into_iter()
            .filter(|r| !removed_repos.contains(&r.name))
            .collect();

        write_sources(remaining_packages, remaining_repos)?;
    }

    if had_errors {
        return Err(Error::Other("Some items could not be removed".to_string()));
    }

    Ok(())
}
